lib/topological_inventory/orchestrator/event_manager.rb

```diff
@@ -0,0 +1,126 @@
+module TopologicalInventory
+  module Orchestrator
+    # Event Manager subscribes to kafka topic "platform.sources.event-stream"
+    # and listens to messages from Sources API: create/update/destroy
+    # on models used for computing of digest (see Source#digest_values)
+    #
+    # When event is received then sync API -> OpenShift is invoked.
+    # Sync is invoked also once per hour
+    class EventManager
+      include Logging
+
+      SCHEDULED_EVENT_INTERVAL = 1.hour
+      POLL_TIME = 10.seconds
+
+      def self.run!(worker)
+        manager = new(worker)
+        manager.run!
+      end
+
+      def initialize(worker)
+        self.worker = worker
+        self.sync_semaphore = Mutex.new
+        self.queue = Queue.new
+      end
+
+      def run!
+        Thread.new { event_listener }
+        Thread.new { scheduler }
+
+        processor
+      end
+
+      private
+
+      attr_accessor :queue, :sync_semaphore, :worker
+      #
+      # Event listener invokes sync when received event from Sources API
+      #
+      def event_listener
+        messaging_client.subscribe_topic(subscribe_opts) do |message|
+          if events.include?(message.message)
+            queue.push(:event_name => message.message,
+                       :model_id   => message.payload['id'])
+          end
+        end
+      ensure
+        messaging_client&.close
+      end
+
+      #
+      # Scheduler invokes sync once per hour
+      #
+      def scheduler
+        loop do
+          queue.push(:event_name => "Scheduled.Sync",
+                     :model_id   => nil)
+
+          sleep(SCHEDULED_EVENT_INTERVAL)
+        end
+      end
+
+      #
+      # Processor starts sync once per 10 seconds
+      #   if there is an event in the queue
+      #
+      def processor
+        loop do
+          events = []
+          events << queue.pop until queue.empty?
+
+          # Sources UI (through API) generates multiple events at the same time
+          # One sync is sufficient
+          if (event = events.first).present?
+            process_event(event[:event_name], event[:model_id])
+          end
+
+          sleep(POLL_TIME)
+        end
+      end
+
+      def process_event(event_name, model_id = nil)
+        # Just one sync at the same time
+        sync_semaphore.synchronize do
```

Comment:
Now that process_event is only called from one place we don't need this lock anymore?

Reply:
right, I overcombined it, just a moment...